### PR TITLE
New deserializer

### PR DESCRIPTION
**Pull request by jareguo**
Reopen from #7251

> Re: https://github.com/cocos-creator/3d-tasks/issues/3763
>
> Changes:
>  * see https://github.com/cocos-creator/engine/pull/6645

Could be safely merged since I submitted to a new branch.

### CODE REVIEW

File: scripts/build-engine/src/index.ts [hunk 1]
```diff
@@ -307,8 +307,14 @@ async function _doBuild ({
...
         rollupPlugins.push(rpTerser({
+            // see https://github.com/terser/terser#compress-options
             compress: {
-                reduce_funcs: false // reduce_funcs not suitable for ammo.js
+                reduce_funcs: false, // reduce_funcs not suitable for ammo.js
+                keep_fargs: false,
+                unsafe_Function: true,
+                unsafe_math: true,
+                unsafe_methods: true,
+                passes: 2,  // first: remove deadcodes and const objects, second: drop variables
```
Comment (jareguo) [inline scripts/build-engine/src/index.ts:353]: 新增了一个 pass（和 2D 一样）
第一遍负责把 const obj = { ONE: 1 } 编译成 var ONE = 1
第二遍负责把 ONE 直接内联到调用的地方
而且这样整个引擎其它地方的包体也会进一步减小。

File: package.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
   "devDependencies": {
     "@babel/core": "^7.9.0",
     "@babel/preset-env": "7.8.7",
-    "@cocos/babel-preset-cc": "2.1.0",
+    "@cocos/babel-preset-cc": "^2.2.0",
```
Comment (shrinktofit) [inline package.json]: I suggest we using exact version here.

File: scripts/build-engine/package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "@babel/plugin-transform-classes": "^7.8.6",
     "@babel/plugin-transform-for-of": "^7.9.0",
     "@babel/preset-env": "^7.8.7",
-    "@cocos/babel-preset-cc": "2.0.0",
```
Comment (shrinktofit) [inline scripts/build-engine/package.json:11]: Again.

File: tsconfig.json [hunk 4]
```diff
@@ -7,6 +7,7 @@
       "outDir": "./bin/.ts",
       "rootDir": "./",
       "strict": true,
+      "strictFunctionTypes": false,
```
Comment (shrinktofit) [inline tsconfig.json]: Why do you need to toggle this strict checking off?

File: cocos/core/utils/js-typed.ts [hunk 5]
```diff
@@ -30,6 +30,16 @@ export function isString (object: any) {
     return typeof object === 'string' || object instanceof String;
 }
 
+/**
+ * Checks whether the obj is empty
+ */
+export function isEmptyObject (obj: any) {
```
Comment (shrinktofit) [inline cocos/core/utils/js-typed.ts:39]:
```ts
/**
 * Checks if the object `obj` does not have one or more enumerable properties(including properties from proto chain).
 * @param obj The object.
 * @returns The result. Note that if the `obj` is not of type `'object'`, `true` is returned.
 */
```

File: cocos/core/assets/texture-base.ts [hunk 6]
```diff
@@ -277,9 +278,11 @@ export class TextureBase extends Asset {
      * @return
      */
     public _serialize (exporting?: any): any {
-        return this._minFilter + ',' + this._magFilter + ',' +
-            this._wrapS + ',' + this._wrapT + ',' +
-            this._mipFilter + ',' + this._anisotropy;
+        if (EDITOR || TEST) {
```
Comment (shrinktofit) [inline cocos/core/assets/texture-base.ts:296]: Why?

File: cocos/core/data/deserialize.ts [hunk 7]
```diff
@@ -28,800 +27,957 @@
...
+    (obj: Mat4, data: number[]) => {
+        Mat4.fromArray(obj, data, 1);
+    }
+];
+
+function serializeBuiltinValueTypes (obj: ValueType): IValueTypeData | null {
+    const ctor = obj.constructor as typeof ValueType;
+    const typeId = BuiltinValueTypes.indexOf(ctor);
+    switch (ctor) {
+        case Vec2:
```
Comment (shrinktofit) [inline cocos/core/data/deserialize.ts:101]: The coupling here can be optimized. Let's discuss.

File: tsconfig.json (hunk 4 above)
Comment (jareguo) [reply tsconfig.json]: This is a stupid check, I need to assign:
```ts
const BuiltinValueTypeSetters: Array<((obj: ValueType, data: number[]) => void)> = [
    ...    
    (obj: Vec3, data: number[]) => {
        obj.x = data[1];
        obj.y = data[2];
        obj.z = data[3];
    }
    ...
];
```
IMO, Vec3 should be able to safely assign to ValueType.

File: cocos/core/assets/texture-base.ts (hunk 6 above)
Comment (jareguo) [reply cocos/core/assets/texture-base.ts:296]: Remove serializations from runtime, reducing engine size.

File: tsconfig.json (hunk 4 above)
Comment (shrinktofit) [reply tsconfig.json]: I think this is reasonable.

You declared `BuiltinValueTypeSetters` as `Array<((obj: ValueType, data: number[]) => void)>`. In any other source portion, it's valid to write `BuiltinValueTypeSetters[0](new ValueType(), ...)` even the setter at `[0]` should accept `Vec3`. So it's prohibited by tsc.

----------------------------

I have summited a PR for you to solve this problem: https://github.com/jareguo/engine/pull/4 and revert the `strictFunctionTypes`.

File: package.json [hunk 8]
```diff
@@ -26,7 +26,7 @@
     "@babel/core": "^7.9.0",
     "@babel/preset-env": "7.8.7",
     "@cocos/babel-preset-cc": "2.2.0",
-    "@cocos/build-engine": "4.0.0-jsb.alpha.12",
+    "@cocos/build-engine": "4.0.0-jsb.alpha.13",
```
Comment (jareguo) [inline package.json:29]: 需要更新 package-lock.json

File: scripts/build-engine/package-lock.json [hunk 9]
```diff
@@ -1146,16 +1290,18 @@
       }
     },
     "@cocos/babel-preset-cc": {
-      "version": "2.0.0",
-      "resolved": "https://registry.npm.taobao.org/@cocos/babel-preset-cc/download/@cocos/babel-preset-cc-2.0.0.tgz",
-      "integrity": "sha1-8QVeY4TtsDTcqgpX7HWbvwW+7jI=",
+      "version": "2.2.0",
```
Comment (jareguo) [inline scripts/build-engine/package-lock.json:1293]: @shrinktofit 需要发布一下 build-engine 的最新版，之前 package-lock.json 里面的 babel-preset-cc 版本还是旧的，我升级上去了

**Pull request merged**